MichaelFedora/simplebar
# feat: isomorphic testing for vue 2 and 3 and relevant fixes

## TL;DR
Picks up on initial work made by @MichaelFedora for https://github.com/Grsmto/simplebar/pull/609 and add a couple of tweaks to make it fully cross-compatible with vue2 & 3.  The bulk of work was related to testing, to ensure that the same code could work on both vue versions and a couple of extra fixes.

### Testing
- I've done isomorphic vue testing once before and it was a PITA, but this one completely over the roof. What a nightmare.
- adds [swap-vue.js](https://github.com/vuelidate/vuelidate/blob/c17ad8927417c71378099c06ec914b4692c4c521/scripts/swap-vue.js) from `@vuelidate` combined with `vue-demi`: this is what allows us to correctly and dynamically synchronize `vue` core library version with the required dependencies (mostly `vue-template-compiler`)
- the current "vue context" is set via npm script from `package.json` —  `yarn use-vue:3` will set context for vue 3 version — so `yarn use-vue:3 && yarn test` will run tests using vue3. There's also `use-vue:2` and `use-vue:2.7`.
    - For convenience ~~`yarn test:all`~~ `yarn test` will run tests for all 3 vue versions sequentially, so that the CI can run it out-of-the-box
- jest config needs to dynamically select `vue-jest` version based on the current context vue version: `vue-jest@4 => vue@2 | vue-jest@5 => vue@3`[^1]. Took the opportunity and cleaned up the "jest" config entry from `package.json` to avoid confusions on what is the source of truth
- `@vue-test/test-utils` also needs to be dynamically `required` because v2 only supports vue3 and v1 for vue2. The library is then re-exported via `./test-utils.js` file so when writing tests, from now one use this:
    ```diff
    - import { mount } from '@vue/test-utils'
    + import { mount } from './test-utils' // to get the correct version for the current vue
    ```
- Found a caveat with `jest-serializer-vue` for vue3 so you have to call `.html()` before `.toMatchSnapshot()`. I've made a PR to the repo, but don't expect for it to be merge as package seems abandoned. https://github.com/eddyerburgh/jest-serializer-vue/pull/56


[^1]: Because we use jest@24! If we migrate that, we need to update [vue-jest ](https://github.com/vuejs/vue-jest#installation)dependency accordingly!!

### Fixes    
- add `engines: { node: <16 }` to `package.json`: not sure if related with the fact that I'm on a m1 mac, but I could not setup the project because of some `sharp` [native bindings](https://github.com/lovell/sharp/issues/1882) using `node@16`. Using nvm tested `node@12` and `node@14` and both work ok. That being said, the base branch is outdated so it might no be relevant when syncing with `master`.
- [[6205851](https://github.com/MichaelFedora/simplebar/pull/1/commits/6205851505bc1d4b99de916aa994b445f2850d92)] recovers `element` `$ref`, and how attributes were parsed — now that have tests, we can see it was failing when using `this.$attrs`.
- declared scroll event via `emits` option and added spec asserting it
- fix a memory/event leak since Simplebar instance was not being destroyed and as such conditional rendering or simply navigating back and forth to a route that had a Simplebar component would leak event listeners. This can be considered out-of-scope so I can remove it from this Pr if desired.
